[WIP] Feature/math guide (#192)
stan-dev/example-models: knitr/math-guide/math_guide.Rmd

---
title: "Current State of Writing Functions in Stan-Math"
author: "Stan Development Team"
date: "24 January 2021"
output: html_document
---

```{r setup, include=FALSE}
library("cmdstanr")
```

This is meant as a basic guide for writing and maintaining functions in the
[stan-dev/math](https://github.com/stan-dev/math) repository. This is the
automatic differentiation (autodiff) library behind the Stan language, and
the vast majority of the functions exposed at the Stan language level are
implemented here in C++.

In the course of the Math library's existance, C++ has changed substantially.
Math was originally written before C++11. It currently targets C++14. In the near
future it will transition to C++17. With this in mind, there are many different
ways to write Math functions. This guide tries to document best practices,
conventions which not all functions in Math follow, but should be followed for
new code to keep the code from getting unweildy (the old patterns will be
updated eventually).

The title contans "Current State" to emphasize that if any information here is
out of date or any advice does not work, it should be reported as a bug (the

> [example-models](https://github.com/stan-dev/example-models)).

Reviewer: [todo] move this over to the math site.

Similarly, the C++ signatures `double sin(const double&)`, or
`double sin(const double)` would also work. For brevity, a discussion on the
const-ness of arguments is deferred to later and simple pass-by-copy signatures
are discussed for now.

There are to functions above because `real` in Stan can correspond to either a
`double` or a `var`. The simplest way to implement this function would be to
define in `prim/fun` the generic signature:

```cpp
template <typename T>
T sin(T);
```

Assuming the implementation is compatible with both `double` and `var`, this is
all that is needed.

A function `real norm(vector)` similarly requires the following C++ functions
to be defined:

```cpp
double norm(Eigen::VectorXd);
var norm(Eigen::Matrix<var, Eigen::Dynamic, 1>);
```

In this case, there is no generic `T norm(T)` function to write because the
input argument is actually different than output type. To handle this, Stan
provides the template metaprogram `return_type_t`. For a given template
argument `T`, `return_type_t` computes the scalar autodiff return type.

- If `T` is `double`, `return_type_t<T>` is `double`
- If `T` is `var`, `return_type_t<T>` is `double`

> - If `T` is `Eigen::Matrix<var, Eigen::Dynamic, 1>`, `return_type_t<T>` is `double`

Reviewer 2: Should return_type_t<T> return var in lines 155 and 156?

## Prim (fully templated)

### Exceptions

Math functions are expected to validate inputs and throw exceptions.
It should be impossible to cause a segfault by calling a
user-facing Math function or trigger an exception that is not handled
by Stan itself. This means that argument sizes should be checked to be
compatible. For instance, a matrix multiply function would need to check
that the first matrix can in fact be multiplied by the second, if this
is not so it should throw.

If a scalar argument should be positive, an exception should be thrown
if it is not. If a matrix argument should be positive definite, the
function should check this before operating on the matrix. This can
lead to significant overhead in the checks, but the current library
policy is to get informative errors back to the user, which means doing
these checks.

Error checks should be done with the functions in
[stan/math/prim/err](https://github.com/stan-dev/math/tree/develop/stan/math/prim/err)
which throw consistently formatted errors.

### Reference types

Any time a function takes a vector or matrix type, it needs to be able
to handle an expression. Some expressions can be expensive to evaluate, so

> when they are evaluated, they should be saved. This includes any arithmetic

Reviewer 3: ```suggestion
each expression should be evaluated only once. If the result is needed multiple times it should be saved. Expressions include any arithmetic
```

operation, a matrix multiply, or a sum, or something else. Some expressions

> are cheap to evaluate though, any of the Eigen views quality here (transpose,

Reviewer 3: ```suggestion
are cheap to evaluate though, any of the Eigen views qualify here (transpose,
```

or matrix variable `x`, `to_ref(x)` returns an Eigen reference to the
input variable `x`. If `x` was an expensive expression, it will be evaluated.
If it was a cheap expression, the reference type won't evaluate. If it
isn't an Eigen type, the `to_ref` just forwards.

```cpp
template <typename T,
          require_eigen_t<T>* = nullptr>
auto myfunc(const T& x) {
  const auto& x_ref = to_ref(x);
}
```

### Holders

If a function returns an Eigen expression, it may be necessary to use
the `make_holder` utility.

Returning expressions is tricky because the expression may have a longer
lifetime than the objects it operates on. For instance, returning an
expression on a local variable leads to this situation:

```cpp
template <typename T>
auto add_zero(const T& a) {
  Eigen::VectorXd b = Eigen::VectorXd(a.size());
  return a + b;
}
```

Because the return type of this function is `auto`, it is possible
that it returns an Eigen expression.

The following code will segfault on the construction of `c` because
by this time the temporary `b` will have been destructed.

```
Eigen::VectorXd a(5);
Eigen::VectorXd c = add_zero(a);
```

> `make_holder` can be used to extend the scope of `b` so that it does
> not immediately get destructed.

Reviewer 3: ```suggestion
`make_holder` can be used to extend the lifetime of `b` so that it matches the expression it is used in and does
not immediately get destructed.
```

```cpp
template <typename T>
auto add_zero(const T& a) {
  Eigen::VectorXd b = Eigen::VectorXd(a.size());

>   return make_holder([](const auto& l, const auto& r) { return l + r; }, std::move(b));

Reviewer 3: This would not work. The functor accepts two arguments and you are only passing one into `make_holder`.

[here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/meta/holder.hpp).

### Tests

Tests in Math are written with the Google test framework. At this
point there are enough functions in Math that the way to get started
is by just copying the tests of a similar Function and editing them to
suit the new function. There are two basic sets of functions that every
Stan function should have. First, for a function named `foo`, there
should be a file in `test/unit/math/prim/fun` named `foo_test.cpp`.

This should include tests that:

1. Check that every exception is throwing in the right way at least once
and that all edge cases are handled correctly.

  For instance, if there is a `check_size_match` function, then there
  should be a test that makes sure that if the sizes don't match, this
  error gets thrown. It is not necessary to check the error message in
  the tests, just check that an exception is thrown and the exception
  type is correct.

2. Check that the function returns the expect values for a couple
suitable inputs.

  For some functions these checks should be more extensive that others,
  but there should at least be a two or three even in simple cases.

> The `prim` tests should only have scalar arithmetic types. No autodiff

Reviewer 3: ```suggestion
The `prim` tests should only have arithmetic types. No autodiff
```

values. The second set of tests should be in `stan/math/mix/fun` with
the same name `foo_test.cpp`. These tests will be used to test all
forms of the autodiff.

The tests in `mix/fun` should use the `expect_ad` function to test
all forms of autodiff. `expect_ad` takes as its first argument a
function to test and then up to three other arguments, all with
arithmetic scalar types. For a given function `f`, and two arguments
`a` and `b`, `expect_ad` tests that:

1. For every combination of `a` and `b`, the returned value of `f`
matches the returned value with non-autodiff types.

2. The finite difference gradients of `f` match those computed with
autodiff.

3. Any time a `f` throws with arithmetic arguments, it also throws
with autodiff arguments.

## Reverse Mode

### Arena memory

Reverse mode autodiff in Math requires a huge number of temporaries and
intermediates to be computed and saved for the reverse pass. There are
so many allocations that the overhead of `malloc` becomes noticable. To
avoid this, Math provides its own memory arena. The assumptions of
the Math arena are:

1. Individual objects allocated in the arena do not need their destructors

> called

Reviewer 3: ```suggestion
called (although there are exceptions)
```

2. All objects allocated on the arena have the same lifetime (that is the

> lifetime of the stack)

Reviewer 3: ```suggestion
lifetime of the stack - until `recover_memory()` is called)
```

Because of this, the Math arena is very easy to implement and is very
high performance (it is a `malloc` that never needs to free).

Functions that use reverse mode autodiff are allowed to save whatever they
need to the arena. Objects saved here will be available during the reverse
pass.

There is a utility for saving objects into the arena that do need

> their destructors called. This is discussed undere `make_callback_ptr`.

Reviewer 3: ```suggestion
their destructors called. This is discussed under `make_callback_var`.
```

The simplest way to do this is add a `require_any_st_var<...>` to
the reverse mode implementation and a `require_all_not_st_var<...>` to
the prim implementation.

### Values

`value_of(x)` returns the values of the variable `x`. If `x` is not an
autodiff type, it simply forwards `x` along. The values of `x` have
the same shape as `x`. For reverse mode autodiff, values are `double`
variables. For higher order autodiff (`fvar<var>`, etc.) this is not
always the case. See `value_of_rec` for a different behavior.

### Values and adjoint extensions to Eigen

The AOS matrix and vector autodiff types come with an extra `.val()`,
`.val_op()`, `.adj()`, and `.adj_op()` member functions.

`.val()` and `.val_op()` return expressions that evaluate to the values
of the autodiff matrix.

`.adj()` and `.adj_op()` return expressions that evaluate to the adjoints
of the autodiff matrix.

The non-`_op` versions of the functions should be preferred. If the
code does not compile (usually an error about const-ness) try the `_op`
versions. This can happen when multiplying the values/adjoints of an AOS
autodiff type against other matrices.

> It is not safe to use `.noalias()` with the AOS value/adjoint expressions.

Reviewer 3: Why not?

Author: I'm not sure, but I believe it exploded when I did it before. When @SteveBronder did this he added some if statements so the noalias code only ran for SOA. I forget where it was though (I thought multiply but when I looked it wasn't there).

These functions also work for SOA types, but return arena types instead.

### Implementing the reverse pass

#### `reverse_pass_callback`

`reverse_pass_callback` is a function for adding code to the reverse pass.
Calling `reverse_pass_callback` with a functor `f` creates an object on the
reverse mode stack that will call `f` when it is chain'd.

#### `make_callback_var`

`make_callback_var(x, f)` is similar to `reverse_pass_callback`, but it constructs
an SOA autodiff type from the argument `x`. The function `f` is called on the
reverse pass similarly to `reverse_pass_callback`.

### Multiple argument scalar types

> Pretend for a minute that 

Reviewer 3: This looks unfinished.

### Tests (SOA types)

SOA tests are separate from the regular AOS tests. To test SOA types, add for
every `expect_ad` call an `expect_ad_matvar` call. `expect_ad` builds on the
`prim` tests by using finite derivatives to check all of the higher order
autodiff. `expect_ad_matvar` builds on the `expect_ad` tests by checking the
SOA autodiff agains the AOS autodiff.

## Pitfalls

### Copying non-arena variables to lambdas used in the reverse pass

### Returning arena types

### `.val()` vs. `.val_op()`

### Returning expressions

### Const correctness, reverse mode autodiff, and arena types

## Handy tricks

### `make_callback_ptr`

### `forward_as`

### Nested Stacks

### Scoped Stacks

As an aside, in order to get the actual vector result, Eigen defines the
`.eval()` operator, so `c.eval()` would return an actual `Eigen::VectorXd`.
Similarly, Math defines the `eval` function which will evaluate an Eigen

> expression and forward anything that is not an Eigen expression.

Reviewer 3: This seems to be in wrong section.
